man/nm-online.1.in
.\" nm-online (1) manual page
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public Licence along
.\" with this manual; if not, write to the Free Software Foundation, Inc.,
.\" 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
.\"
.\" Copyright (C) 2010 - 2014 Red Hat, Inc.
.\"
.TH NM-ONLINE "1" "2014-03-25" "NetworkManager 1.2"

.SH NAME
nm-online \- ask NetworkManager whether the network is connected
.SH SYNOPSIS
.B nm-online
.RI " [ " OPTIONS " ]

.SH DESCRIPTION
.B nm-online
is a utility to find out whether we are online. It is done by asking
NetworkManager about its status. When run, \fInm\-online\fP waits until
NetworkManager reports an active connection, or specified timeout expires. On
exit, the returned status code should be checked (see the return codes below).
.P
By default NetworkManager waits for IPv4 dynamic addressing to complete but does
not wait for the "auto" IPv6 dynamic addressing.  To wait for IPv6 addressing to
complete, either (1) change the network connection's IPv6 "may-fail" setting to "no",
and/or (2) change the IPv6 addressing method to "manual" or "dhcp", to indicate that
IPv6 connectivity is expected.
.SH OPTIONS
.TP
.B \-t, \-\-timeout <timeout_value>
Time to wait for a connection, in seconds.  If the option is not provided,
the default timeout is 30 seconds.
.TP
.B \-x, \-\-exit
Exit immediately if NetworkManager is not running or connecting.
.TP
.B \-q, \-\-quiet
Don't print anything.
.TP
.B \-s, \-\-wait-for-startup
Wait for NetworkManager startup to complete, rather than waiting for
network connectivity specifically. Startup is considered complete once
NetworkManager has activated (or attempted to activate) every
auto-activate connection which is available given the current network
state. (This is generally only useful at boot time; after startup has
completed, \fInm-online -s\fP will just return immediately, regardless
of the current network state.)

.SH EXIT STATUS

.IP "0" 4
Success - already online or connection established within given timeout
.IP "1" 4
Offline or not online within given timeout
.IP "2" 4
Unspecified error

.SH SEE ALSO
.BR nmcli (1),
.BR NetworkManager(8).


NEWS
===============================================
NetworkManager-1.0.12
Overview of changes since NetworkManager-1.0.10
===============================================

* DHCP leases on software devices are now renewed when the computer is
  awoken from suspend.

* Improved ifupdown plugin robustness and interoperability on Yocto and
  OpenEmbedded.

* Fixed failed VPN activations when plugin supports interactive mode, but
  the VPN daemon does not.

* Wi-Fi monitor interfaces are now ignored, not turned into managed mode.

* AP and AdHoc mode connections with manual IP configuration are now able
  to autoconnect.

* Broken device drivers (AWS ENI) that initially have invalid MAC addresses
  are now properly managed as soon as correct MAC address is set.

* WWAN devices are unlocked a bit earlier so that supported IP versions can
  be queried.

* The NetworkManager.service was ordered after network-pre.target and
  dbus.service. This ensures NetworkManager doesn't set up connectivity
  before firewall rules are in place and wouldn't exit before remote
  filesystems can be umounted ensuring orderly operation of systemd
  managed installations.

* The netfilter rules used with shared IPv4 method are now removed on exit.

* Ability to manage USB gadget drivers (UDC side) has beed added.

* Infiniband transport mode change now takes place with the link set down
  because some drivers need that.

* Race conditions that could disclose connection secrets to authenticated
  local users when changing ifcfg and keyfile connections have been fixed.
  This has security impact of low severity. (CVE-2016-0764)

* A handful of memory leak and crasher bugs of minor importance have been
  fixed.

===============================================
NetworkManager-1.0.10
Overview of changes since NetworkManager-1.0.8
===============================================

This is a new stable release of NetworkManager.  Notable changes include:

* Added support for handling VPN secrets to nmtui and nmcli agent.

* Fixed a regression that caused NetworkManager to ignore external deletion of
  a device with master.

* Fixed glitches with older versions of glib (prior to 2.36.0).

* Fixed build with most recent versions of libsoup.

* Fixed busy retry loop on non-transient errors from wpa_supplicant.

* Improvements to testing infrastructure.

* Updates to German, Galician, Occitan, Turkish, Ukrainian, Brazilian, Swedish,
  Polish and Italian translations.

===============================================
NetworkManager-1.0.8
Overview of changes since NetworkManager-1.0.6
===============================================

This is a new stable release of NetworkManager.  Notable changes include:

* MTU indicated by a VPN gateway is now properly applied.

* Fixed MSS setting when MTU changes.

* The default route is properly restored on device disconnect.

* Build with older toolchains has been fixed.

* The team devices can now properly be enslaved to bridges.

* Failed DHCP attempts for assumed connections are now retried after
  a timeout.

* Default wired connection is now created after udev registers the device.

* Support for Bluetooth DUN devices with Bluez 5 has been fixed.

* The ipv6.ignore-auto-dns property is now properly honored making it
  possible to override automatically obtained name servers.

* Invalid permanent MAC adddresses as reported by some devices are now
  ignored.

* Device links reported by more recent versions of Linux kernel that reside
  in different network namespaces are no longer confused with links in
  the namespace NetworkManager runs in.

* MAC address changes of VLANs enslaved to a bond are now properly
  propagated to the master device.

* Fixed error handling for teaming devices with invalid configuration.

* Wi-Fi AP list is now updated correctly after AP mode has been used.

* Management of a device is not attempted until the device has been
  registered with udev.

* The error handling for VPN secret agents is now significantly more robust.

* Detection of s390 CTC devices now works properly.

* A GATEWAY property in /etc/sysconfig/network now no longer affects
  non-static connections.

* Added support for IPv6-only VPN connections.

* The systemd service now uses HUP signal to reload configuration.

* Change VLAN default flags to set REORDER_HDR for new connections.

* nmtui is now able to ignore automatically configured routes.

* Numerous bash shell autocompletion fixes for nmcli.

* Allow setting IPv6 and PPP settings for GDM and CDMA connections via
  nmcli.

* Added support for adding ADSL connections in nmcli.

* Numerous crash fixes.

* Updates to Swedish, Italian and Polish translations.

===============================================
NetworkManager-1.0.6
Overview of changes since NetworkManager-1.0.4
===============================================

This is a new stable release of NetworkManager.  Notable changes include:

* Improved capture portal detection.

* Default route through WiFi connection is now preferred to Mobile
    Broadband if both are available.

* Expose a flag to determine whether a particular connection is metered
    via API and client tools.

* Add support for locking connections to a channel within a particular
    band.

* Add support for configuring Wake-on-LAN capabilitites.

* Allow overriding the MTU for team device.

* Usual pile of bug fixes and robustness improvements.

===============================================
NetworkManager-1.0.4
Overview of changes since NetworkManager-1.0.2
===============================================

This is a new stable release of NetworkManager.  Notable changes include:

* The MTU setting from an IPv6 neighbor discovery Router Advertisements is
    now ignored if applying it would result in invalid configuration.

* Some configuration options can now be changed without restarting the
    daemon. Notably, this applies to 'dns', 'connectivity' and
    'ignore-carrier' settings.


clients/nm-online.c
/* -*- Mode: C; tab-width: 4; indent-tabs-mode: t; c-basic-offset: 4 -*- */
/*
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Copyright (C) 2006 - 2008 Novell, Inc.
 * Copyright (C) 2008 - 2014 Red Hat, Inc.
 *
 */

/*
 * nm-online.c - Are we online?
 *
 * Return values:
 *
 * 	0	: already online or connection established within given timeout
 *	1	: offline or not online within given timeout
 *	2	: unspecified error
 *
 * Robert Love <rml@novell.com>
 */

#include "nm-default.h"

#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <locale.h>

#include "NetworkManager.h"

#define PROGRESS_STEPS 15
#define WAIT_STARTUP_TAG "wait-startup"

typedef struct
{
	gint64 start_timestamp_ms;
	gint64 end_timestamp_ms;
	gint64 progress_step_duration;
	gboolean quiet;
} Timeout;

static void
client_properties_changed (GObject *object,
                           GParamSpec *pspec,
                           gpointer loop)
{
	NMClient *client = NM_CLIENT (object);
	NMState state;
	gboolean wait_startup = GPOINTER_TO_UINT (g_object_get_data (object, WAIT_STARTUP_TAG));

	if (!nm_client_get_nm_running (client))
		return;

	if (wait_startup) {
		if (!nm_client_get_startup (client))
			g_main_loop_quit (loop);
	} else {
		state = nm_client_get_state (client);
		if (   state == NM_STATE_CONNECTED_LOCAL
		    || state == NM_STATE_CONNECTED_SITE
		    || state == NM_STATE_CONNECTED_GLOBAL)
			g_main_loop_quit (loop);
	}
}

static gboolean
handle_timeout (gpointer data)
{
	const Timeout *timeout = data;
	const gint64 now = g_get_monotonic_time () / (G_USEC_PER_SEC / 1000);
	gint64 remaining_ms = timeout->end_timestamp_ms - now;
	const gint64 elapsed_ms = now - timeout->start_timestamp_ms;
	int progress_next_step_i = 0;

	if (!timeout->quiet) {
		int i;

		/* calculate the next step (not the current): floor()+1 */
		progress_next_step_i = (elapsed_ms / timeout->progress_step_duration) + 1;
		progress_next_step_i = MIN (progress_next_step_i, PROGRESS_STEPS);

		g_print ("\r%s", _("Connecting"));
		for (i = 0; i < PROGRESS_STEPS; i++)
			putchar (i < progress_next_step_i ? '.' : ' ');
		g_print (" %4lds", (long) (MAX (0, remaining_ms) / 1000));
		fflush (stdout);
	}

	if (remaining_ms <= 3) {
		if (!timeout->quiet)
			g_print ("\n");
		exit (1);
	}

	if (!timeout->quiet) {
		gint64 rem;

		/* synchronize the timeout with the ticking of the seconds. */
		rem = remaining_ms % 1000;
		if (rem <= 3)
			rem = rem + G_USEC_PER_SEC;
		rem = rem + 10; /* add small offset to awake a bit after the second ticks */
		if (remaining_ms > rem)
			remaining_ms = rem;

		/* synchronize the timeout with the steps of the progress bar. */
		rem = (progress_next_step_i * timeout->progress_step_duration) - elapsed_ms;
		if (rem <= 3)
			rem = rem + timeout->progress_step_duration;
		rem = rem + 10; /* add small offset to awake a bit after the time out */
		if (remaining_ms > rem)
			remaining_ms = rem;
	}

	g_timeout_add (remaining_ms, handle_timeout, (void *) timeout);
	return G_SOURCE_REMOVE;
}

int
main (int argc, char *argv[])
{
	int t_secs = 30;
	gboolean exit_no_nm = FALSE;
	gboolean wait_startup = FALSE;
	Timeout timeout;
	GOptionContext *opt_ctx = NULL;
	gboolean success;
	NMClient *client;
	NMState state = NM_STATE_UNKNOWN;
	GMainLoop *loop;
	gint64 remaining_ms;
	GError *error = NULL;

	GOptionEntry options[] = {
		{"timeout", 't', 0, G_OPTION_ARG_INT, &t_secs, N_("Time to wait for a connection, in seconds (without the option, default value is 30)"), "<timeout>"},
		{"exit", 'x', 0, G_OPTION_ARG_NONE, &exit_no_nm, N_("Exit immediately if NetworkManager is not running or connecting"), NULL},
		{"quiet", 'q', 0, G_OPTION_ARG_NONE, &timeout.quiet, N_("Don't print anything"), NULL},
		{"wait-for-startup", 's', 0, G_OPTION_ARG_NONE, &wait_startup, N_("Wait for NetworkManager startup instead of a connection"), NULL},
		{NULL}
	};

	timeout.start_timestamp_ms = g_get_monotonic_time () / (G_USEC_PER_SEC / 1000);
	timeout.quiet = FALSE;

	/* Set locale to be able to use environment variables */
	setlocale (LC_ALL, "");

	bindtextdomain (GETTEXT_PACKAGE, NMLOCALEDIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);

	opt_ctx = g_option_context_new (NULL);
	g_option_context_set_translation_domain (opt_ctx, GETTEXT_PACKAGE);
	g_option_context_set_ignore_unknown_options (opt_ctx, FALSE);
	g_option_context_set_help_enabled (opt_ctx, TRUE);
	g_option_context_add_main_entries (opt_ctx, options, NULL);

	g_option_context_set_summary (opt_ctx,
	                              _("Waits for NetworkManager to finish activating startup network connections."));

	success = g_option_context_parse (opt_ctx, &argc, &argv, NULL);
	g_option_context_free (opt_ctx);

	if (!success) {
		g_printerr ("%s: %s\n", argv[0],
		            _("Invalid option.  Please use --help to see a list of valid options."));
		return 2;
	}

	if (t_secs < 0 || t_secs > 3600)  {
		g_printerr ("%s: %s\n", argv[0],
		            _("Invalid option.  Please use --help to see a list of valid options."));
		return 2;
	}
	remaining_ms = t_secs * 1000;

	nm_g_type_init ();

	client = nm_client_new (NULL, &error);
	if (!client) {
		g_printerr (_("Error: Could not create NMClient object: %s."), error->message);
		g_error_free (error);
		return 2;
	}

	loop = g_main_loop_new (NULL, FALSE);

	g_object_set_data (G_OBJECT (client), WAIT_STARTUP_TAG, GUINT_TO_POINTER (wait_startup));
	state = nm_client_get_state (client);
	if (!nm_client_get_nm_running (client)) {
		if (exit_no_nm) {
			g_object_unref (client);
			return 1;
		}
	} else if (wait_startup) {
		if (!nm_client_get_startup (client)) {
			g_object_unref (client);
			return 0;
		}
	} else {
		if (   state == NM_STATE_CONNECTED_LOCAL
		    || state == NM_STATE_CONNECTED_SITE
		    || state == NM_STATE_CONNECTED_GLOBAL) {
			g_object_unref (client);
			return 0;
		}
	}
	if (exit_no_nm && (state != NM_STATE_CONNECTING)) {
		g_object_unref (client);
		return 1;
	}

	if (remaining_ms == 0) {
		g_object_unref (client);
		return 1;
	}

	g_signal_connect (client, "notify",
	                  G_CALLBACK (client_properties_changed), loop);

	timeout.end_timestamp_ms = timeout.start_timestamp_ms + remaining_ms;
	timeout.progress_step_duration = (timeout.end_timestamp_ms - timeout.start_timestamp_ms + PROGRESS_STEPS/2) / PROGRESS_STEPS;

	g_timeout_add (timeout.quiet ? remaining_ms : 0,
	               handle_timeout, &timeout);

	g_main_loop_run (loop);
	g_main_loop_unref (loop);

	g_object_unref (client);

	return 0;
}
